File: debop4s-timeperiod/src/test/scala/com/github/debop4s/timeperiod/tests/timeranges/WeekRangeTest.scala
package com.github.debop4s.timeperiod.tests.timeranges


import com.github.debop4s.timeperiod._
import com.github.debop4s.timeperiod.tests.AbstractTimePeriodTest
import com.github.debop4s.timeperiod.timerange.WeekRange
import com.github.debop4s.timeperiod.utils.Times

/**
 * com.github.debop4s.timeperiod.tests.timeranges.WeekRangeTest
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2014. 2. 17. 오전 9:51
 */
class WeekRangeTest extends AbstractTimePeriodTest {

    test("init values") {
        val now = Times.now
        val firstWeek = Times.startTimeOfWeek(now)
        val secondWeek = firstWeek + 1.week

        val wr = WeekRange(now, EmptyOffsetTimeCalendar)
        wr.start should equal(firstWeek)
        wr.end should equal(secondWeek)
    }

    test("default calendar") {
        val yearStart = Times.startTimeOfYear(Times.now)

        (1 until 50).par.foreach(w => {
            val wr = WeekRange(yearStart + w.week)
            wr.weekyear should equal(yearStart.getWeekyear)
            wr.weekOfWeekyear should equal(w + 1)

            wr.unmappedStart should equal(Times.startTimeOfWeek(yearStart + w.week))
            wr.unmappedEnd should equal(Times.startTimeOfWeek(yearStart + (w + 1).week))
        })
    }

    test("getDays") {
        val now = Times.now
        val wr = WeekRange()

        val days = wr.getDays
        var index = 0
        days.foreach(day => {
            day.start should equal(wr.start + index.day)
            day.end should equal(day.calendar.mapEnd(day.start + 1.day))
            index += 1
        })
        index should equal(DaysPerWeek)
    }

    test("add months") {
        val now = Times.now
        val startWeek = Times.startTimeOfWeek(now)
        val wr = WeekRange(now)

        log.debug(s"startWeek=$startWeek, weekRange=$wr")

        wr.previousWeek.start should equal(startWeek - 1.week)
        wr.nextWeek.start should equal(startWeek + 1.week)
        wr.addWeeks(0) should equal(wr)

        (-60 to 120).par.foreach(w => {
            wr.addWeeks(w).start should equal(startWeek + w.week)
        })
    }
}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/WeekRangeTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.core.JAction1;
import debop4s.core.parallels.JParallels;
import debop4s.timeperiod.TimeCalendar;
import debop4s.timeperiod.TimeSpec;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.DayRange;
import debop4s.timeperiod.timerange.WeekRange;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.timeperiod.test.timeranges.WeekRangeFunSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 5. 26. 오전 12:00
 */
@Slf4j
public class WeekRangeTest extends TimePeriodTestBase {

    @Test
    public void initValues() {
        DateTime now = Times.now();
        DateTime firstWeek = Times.startTimeOfWeek(now);
        DateTime secondWeek = firstWeek.plusWeeks(1);

        WeekRange weekRange = new WeekRange(now, TimeCalendar.getEmptyOffset());

        assertThat(weekRange.getStart()).isEqualTo(firstWeek);
        assertThat(weekRange.getEnd()).isEqualTo(secondWeek);
    }

    @Test
    public void defaultCalendarTest() {
        DateTime yearStart = Times.startTimeOfYear(Times.now());

        for (int w = 1; w < 50; w++) {

            WeekRange weekRange = new WeekRange(yearStart.plusWeeks(w));
            assertThat(weekRange.year()).isEqualTo(yearStart.getYear());
            assertThat(weekRange.weekOfWeekyear()).isEqualTo(w + 1);

            assertThat(weekRange.getUnmappedStart()).isEqualTo(Times.startTimeOfWeek(yearStart.plusWeeks(w)));
            assertThat(weekRange.getUnmappedEnd()).isEqualTo(Times.startTimeOfWeek(yearStart.plusWeeks(w)).plusWeeks(1));
        }
    }


    @Test
    public void getDaysTest() {
        final DateTime now = Times.now();
        final WeekRange weekRange = new WeekRange();
        List<DayRange> days = weekRange.getDays();

        int index = 0;
        for (DayRange day : days) {
            assertThat(day.getStart()).isEqualTo(weekRange.getStart().plusDays(index));
            assertThat(day.getEnd()).isEqualTo(day.getTimeCalendar().mapEnd(day.getStart().plusDays(1)));
            index++;
        }
        assertThat(index).isEqualTo(TimeSpec.DaysPerWeek);
    }

    @Test
    public void addMonthsTest() {

        final DateTime now = Times.now();
        final DateTime startWeek = Times.startTimeOfWeek(now);
        final WeekRange weekRange = new WeekRange(now);

        log.debug("startWeek=[{}], weekRange=[{}]", startWeek, weekRange);

        assertThat(weekRange.previousWeek().getStart()).isEqualTo(startWeek.minusWeeks(1));
        assertThat(weekRange.nextWeek().getStart()).isEqualTo(startWeek.plusWeeks(1));

        assertThat(weekRange.addWeeks(0)).isEqualTo(weekRange);

        JParallels.run(-60, 120, new JAction1<Integer>() {
            @Override
            public void perform(Integer m) {
                assertThat(weekRange.addWeeks(m).getStart()).isEqualTo(startWeek.plusWeeks(m));
            }
        });
    }
}
